### ISSUE TITLE
Define mapping of IoT related information to Cloud Events.

### ISSUE THREAD

**Issue by ctron** [drogue-iot/drogue-cloud#29]
Currently we have a little bit of a mess going on with IoT related information (device id, model id, …), mapping them to the cloud events attributes.

We need to define how to best use the existing attributes, and fix any spec violating constructs that we currently might have.

**Comment by arnab15** [comment]
hello,
I have knowledge of javascript ,nodejs,reactjs,expressjs, Mongodb but I haven't contributed to any open-source project before so I'm a beginner. I would like to contribute to this project.
how should I start can anyone suggest to me the way should I start contributing.

**Comment by garyedwards** [comment]
Google have been similar work with their IoT Core + PubSub which supports their flavour of CloudEvents and Knative.  Some of their efforts around naming and tools can be found in the following projects as a reference:

- https://github.com/faucetsdn/udmi
- https://github.com/faucetsdn/udmi_site_model
- https://github.com/google/digitalbuildings

It may be good to collaborate around some of these standards to make portability between GCP and Drogue easier.  We are playing with both currently so happy to help in the effort.

A basic example of the attributes that come from IoT Core are below.  This an be embellished with extra meta data from the UDMI projects listed above.

```
{
  "message": {
    "attributes": {
      "deviceId": "my-device",
      "deviceNumId": "2838737599123417",
      "deviceRegistryId": "test-reg",
      "deviceRegistryLocation": "europe-west1",
      "projectId": "my-test-project"
      "subFolder": ""
    },
    "data": some-payload,
    "messageId": "1964974633438725",
    "message_id": "1964974631834725",
    "publishTime": "2021-01-29T14:12:05.081Z",
    "publish_time": "2021-01-29T14:12:05.081Z"
  },
  "subscription": "projects/my-test-project/subscriptions/mySubscription"
}
```

**Comment by ctron** [comment]
Thanks for the input. That is super helpful.

That looks like it is more about the actual payload. I think supporting this sounds like a reasonable goal. However there are also other efforts like Eclipse Ditto's payload format, with the Vorto device description. Or the AWS digital twin stuff.

For the first step I was hoping we can be payload agnostic as much as possible. We can always consume or publish data in the format of others. Like for example with the Thing Network endpoint, which again has their own way of dealing with metadata and payload.

As a second step, it would definitely be super helpful to take a look at aligning payload formats.

It makes sense to me to map things like "device ID", "tenant/project/application ID", etc or the "message_id" mentioned above, to some of the existing CloudEvents attributes though. Instead of using "message_id", I think using the "id" makes more sense. Or instead of "publish_time", use the proper "time" attribute of cloud events.

Maybe it is easier to do it from the other side. Assuming we have a bunch of cloud events attribute, what information should protocol adapters/endpoints put in?

If you are willing to help, I guess it would be great if you could go through the list of common cloud event attributes (https://github.com/cloudevents/spec/blob/v1.0.1/spec.md#context-attributes) and let us know what you would like to see in the different attributes.

**Comment by garyedwards** [comment]
True, UDMI is more payload specific but their are parts as demonstrated in the `udmi_site_model` metadata files that would be added to the device registry and then as attributes of events.  I guess the following two docs would be more relevant but are intentionally light on proposing application specific attributes:

- [Kafka Protocol Binding for CloudEvents](https://github.com/cloudevents/spec/blob/v1.0.1/kafka-protocol-binding.md)
- [Google Cloud Pub/Sub Protocol Binding for CloudEvents](https://github.com/google/knative-gcp/blob/master/docs/spec/pubsub-protocol-binding.md)

The `REQUIRED` and `OPTIONAL` attributes seem quite self explanatory and a simple Rosetta Stone could map vendor specific attributes to CloudEvents.

More IoT specific [Attribute Extensions](https://github.com/cloudevents/spec/blob/v1.0.1/documented-extensions.md) may need to be developed to pick up other useful metadata (as per this issue title 😉).  I will have a think and review some existing systems GCP AWS etc.

**Comment by ctron** [comment]
Cool, thanks for your help.

I will go ahead and create a new RFC document for working on this. Please feel free to clone, fork, amend, comment, and/or create PRs (for the PR). So we can have a shared document for working on this.

**Comment by ctron** [comment]
As mentioned on the PR of the RFC: most of the changes should already be implemented. With the exception of the `uid` variant, which we do have present internally.

So while we still need to do some minor work for 0.5, most of the changes should be part of 0.4 already. And everyone should be able to test them out already.

**Comment by ctron** [comment]
I think it is ok to close this for 0.5. We can always make changes in the future.
